cmd/minikube/cmd/start.go

```diff
@@ -500,3 +505,60 @@ func saveConfigToFile(data []byte, file string) error {
 	}
 	return nil
 }
+
+// kubectlVersion is a helper struct used to unmarshal the output of "kubectl version --client -o json".
+type kubectlVersion struct {
+	ClientVersion apimachineryversion.Info `json:"clientVersion,omitempty"`
+}
+
+// checkKubectlCompatibility checks whether kubectl is installed and whether its version is compatible with the desired Kubernetes version.
+func checkKubectlCompatibility(kubernetesVersion string) {
+	// Lookup kubectl in PATH.
+	k, err := exec.LookPath("kubectl")
+	if err != nil {
+		warn("kubectl is not installed or not present in the PATH.")
+		return
+	}
+	// Grab the version reported by kubectl.
+	o, err := exec.Command(k, "version", "--client", "-o", "json").Output()
+	if err != nil {
+		warn("Failed to detect kubectl version.")
+		return
+	}
+	// Parse the version reported by kubectl.
+	v := kubectlVersion{}
+	if err := json.Unmarshal(o, &v); err != nil {
+		warn("Failed to detect kubectl version.")
+		return
+	}
+	svKubectlVersion, err := semver.Parse(strings.TrimPrefix(v.ClientVersion.GitVersion, version.VersionPrefix))
+	if err != nil {
+		warn("Failed to parse the version reported by kubectl.")
+		return
+	}
+
+	// Parse the desired Kubernetes version.
+	svKubernetesVersion, err := semver.Parse(strings.TrimPrefix(kubernetesVersion, version.VersionPrefix))
+	if err != nil {
+		warn("Failed to parse the desired version of Kubernetes.")
+		return
+	}
+
+	// Create a range of allowed versions based on the desired Kubernetes version.
+	svKubernetesRange, err := semver.ParseRange(fmt.Sprintf(">=%d.%d.0 <%d.%d.0", svKubernetesVersion.Major, svKubernetesVersion.Minor-1, svKubernetesVersion.Major, svKubernetesVersion.Minor+2))
+	if err != nil {
+		warn("Failed to create the range of allowed kubectl versions.")
+		return
+	}
+
+	// Warn the user if the version of kubectl doesn't fall within the allowed version range.
+	if !svKubernetesRange(svKubectlVersion) {
+		warn("The version of kubectl installed (%s) is not compatible with the desired Kubernetes version (%s).", svKubectlVersion, svKubernetesVersion)
```

Comment:
I'd rather see this error out by default, and if necessary, we can add a --force flag to allow users to shoot themselves in the foot.

I find that users generally ignore warnings unless they have to correct for them. Incompatible kubectl versions tend to lead to very strange bug reports.

Reply:
Now that I know a bit better - I'm OK with this warning. For consistency, could you make it use console.OutStyle("warning") though?